File: leetcode/Python/partition-array-into-disjoint-intervals.py
class Solution:
    def partitionDisjoint(self, nums: List[int]) -> int:
        mn, mx, t = [0]*len(nums), 0, nums[-1]
        for i in range(len(nums)-1,-1,-1):
            t=min(t,nums[i])
            mn[i] = t
        for i in range(len(nums)-1):
            mx=max(mx,nums[i])
            if mx<=mn[i+1]:
                return i+1
        return len(nums)
        

File: leetcode/C++/partition-array-into-disjoint-intervals.cpp
class Solution {
public:
    int partitionDisjoint(vector<int>& nums) {
        vector<int> mn(nums.size(),0);
        int mx=0, t=nums.back();
        for(int i=nums.size()-1; i>=0; --i){
            t=min(t,nums[i]);
            mn[i]=t;
        }
        for(int i=0; i<nums.size()-1; ++i){
            mx=max(nums[i],mx);
            if(mx<=mn[i+1]) return i+1;
        }
        return nums.size();
    }
};